docs(v8): Restructure migration docs (#10830)
getsentry/sentry-javascript: MIGRATION.md

# Sentry JavaScript SDK Migration Docs

These docs walk through how to migrate our JavaScript SDKs through different major versions.

- Upgrading from [SDK 4.x to 5.x/6.x](./docs/migration/v4-to-v5_v6.md)
- Uprading from [SDK 6.x to 7.x](./docs/migration/v6-to-v7.md)
- Upgrading from [SDK 7.x to 8.x](./MIGRATION.md#upgrading-from-7x-to-8x)

# Upgrading from 7.x to 8.x

The main goal of version 8 is to improve our performance monitoring APIs, integrations API, and ESM support. This
version is breaking because we removed deprecated APIs, restructured npm package contents, and introduced new
dependencies on OpenTelemetry. Below we will outline the steps you need to take to tackle to deprecated methods.

Before updating to `8.x` of the SDK, we recommend upgrading to the latest version of `7.x`. You can then follow
[these steps](./MIGRATION.md#deprecations-in-7x) remove deprecated methods in `7.x` before upgrading to `8.x`.

The v8 version of the JavaScript SDK requires a self-hosted version of Sentry TBD or higher (Will be chosen once first
stable release of `8.x` comes out).

## 1. Version Support changes:

> **Node.js**: We now official support Node 14+ for our CJS package, and Node 18.8+ for our ESM package. This applies to

Reviewer: Actually let's make this `14.8+` as this is what supports async local storage! (I'll also bump the actual engines for this)

`@sentry/node` and all of our node-based server-side sdks (`@sentry/nextjs`, `@sentry/serverless`, etc.). We no longer
test against Node 8, 10, or 12 and cannot guarantee that the SDK will work as expected on these versions.

**Browser**: Our browser SDKs (`@sentry/browser`, `@sentry/react`, `@sentry/vue`, etc.) now require ES6+ compatible
browsers. This means that we no longer support IE11 (end of an era). This also means that the Browser SDK requires the
fetch API to be available in the environment.

New minimum supported browsers:

- Chrome 51
- Edge 15
- Safari/iOS Safari 10
- Firefox 54
- Opera 38
- Samnsung Internet 5

> For IE11 support please transpile your code to ES% using babel or similar and add required polyfills.

Reviewer: ```suggestion
For IE11 support please transpile your code to ES5 using babel or similar and add required polyfills.
```

**React**: We no longer support React 15 in version 8 of the React SDK.

## 2. Package removal

We've removed the following packages:

- [@sentry/hub](./MIGRATION.md#sentryhub)
- [@sentry/tracing](./MIGRATION.md#sentrytracing)
- [@sentry/integrations](./MIGRATION.md#sentryintegrations)

#### @sentry/hub

`@sentry/hub` has been removed and will no longer be published. All of the `@sentry/hub` exports have moved to
`@sentry/core`.

#### @sentry/tracing

`@sentry/tracing` has been removed and will no longer be published. See
[below](./MIGRATION.md/#3-removal-of-deprecated-apis) for more details.

For Browser SDKs you can import `BrowserTracing` from the SDK directly:

```js
// Before
import * as Sentry from '@sentry/browser';
import { BrowserTracing } from '@sentry/tracing';

Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
  integrations: [new BrowserTracing()],
});

// After
import * as Sentry from '@sentry/browser';

Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,

>   integrations: [new Sentry.BrowserTracing()],

Reviewer 2: ```suggestion
  integrations: [new Sentry.browserTracingIntegration()],
```

});
```

If you were importing `@sentry/tracing` for the side effect, you can now use `Sentry.addTracingExtensions()` to add the
tracing extensions to the SDK. `addTracingExtensions` replaces the `addExtensionMethods` method from `@sentry/tracing`.

```js

> // Before

Reviewer 2: super-l: doesn't have to happen now but for clarity, we could label the changes as `v7/v8` vs. `before/after`?

Author: yes good idea!

import * as Sentry from '@sentry/browser';
import '@sentry/tracing';

Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
});

// After
import * as Sentry from '@sentry/browser';

> Sentry.addTracingExtensions();

Reviewer: do we need this if we add browser tracing below?

Author: copy paste error - fixed!

Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,

>   integrations: [new Sentry.BrowserTracing()],

Reviewer: ```suggestion
  integrations: [Sentry.browserTracingIntegration()],
```

Sentry.init({
  integrations: [Sentry.browserTracingIntegration()],
});
```

### `@sentry/react`

```js
import * as Sentry from '@sentry/react';

Sentry.init({
  integrations: [
    // No react router
    Sentry.browserTracingIntegration(),
    // OR, if you are using react router, instead use one of the following:
    Sentry.reactRouterV6BrowserTracingIntegration({
      useEffect,
      useLocation,
      useNavigationType,
      createRoutesFromChildren,
      matchRoutes,
      stripBasename,
    }),
    Sentry.reactRouterV5BrowserTracingIntegration({
      history,
    }),
    Sentry.reactRouterV4BrowserTracingIntegration({
      history,
    }),
    Sentry.reactRouterV3BrowserTracingIntegration({
      history,
      routes,
      match,
    }),
  ],
});
```

### `@sentry/vue`

```js
import * as Sentry from '@sentry/vue';

Sentry.init({
  integrations: [
    Sentry.browserTracingIntegration({
      // pass router in, if applicable
      router,
    }),
  ],
});
```

### `@sentry/angular` & `@sentry/angular-ivy`

```js
import * as Sentry from '@sentry/angular';

Sentry.init({
  integrations: [Sentry.browserTracingIntegration()],
});

> // You still need to add the Trace Service like before!

Reviewer 2: l:
```suggestion
// You still need to add the TraceService like before!
```
